// nutrient-data-laboratory/index.js
'use strict'

const https = require('https')
const querystring = require('querystring')
const url = require('url')

// https://ndb.nal.usda.gov/ndb/doc/index
class NutrientDataLaboratory {
  constructor (apiKey) {
    this.hostname = 'https://api.nal.usda.gov/ndb'
    this.apiKey = apiKey
  }

  static request (hostname, apiKey, path, params, callback) {
    params.api_key = apiKey
    params.format = params.format || 'JSON'

    let qs = querystring.stringify(params)
    let url = `${hostname}/${path}/?${qs}`

    https.get(url, (res) => {
      res.setEncoding('utf8')
      let rawData = ''
      res.on('data', (chunk) => rawData += chunk)
      res.on('end', () => callback(null, JSON.parse(rawData)))
    }).on('error', (err) => callback(err.message))
  }

  // https://ndb.nal.usda.gov/ndb/doc/apilist/API-FOOD-REPORT.md
  foodReports (params, callback) {
    NutrientDataLaboratory.request(
      this.hostname,
      this.apiKey,
      'reports',
      params,
      callback
    )
  }

  // https://ndb.nal.usda.gov/ndb/doc/apilist/API-LIST.md
  lists (params, callback) {
    NutrientDataLaboratory.request(
      this.hostname,
      this.apiKey,
      'list',
      params,
      callback
    )
  }

  // https://ndb.nal.usda.gov/ndb/doc/apilist/API-NUTRIENT-REPORT.md
  nutrientReports (params, callback) {
    NutrientDataLaboratory.request(
      this.hostname,
      this.apiKey,
      'nutrients',
      params,
      callback
    )
  }

  // https://ndb.nal.usda.gov/ndb/doc/apilist/API-SEARCH.md
  search (params, callback) {
    NutrientDataLaboratory.request(
      this.hostname,
      this.apiKey,
      'search',
      params,
      callback
    )
  }
}

module.exports = NutrientDataLaboratory
